src/hooks/useConnectWallet.ts

```diff
@@ -1,49 +1,201 @@
-    const nextIndex = currentIndex < accounts.length - 1 ? currentIndex + 1 : 0
-
-    const nextAccountId = accounts[nextIndex].accountId
+  const handleSignOutViaWalletConnect = async () => {
+    await disconnect()
+  }
 
-    selector.setActiveAccount(nextAccountId)
+  const abstractExecutor =
+    (
+      executors: Record<
+        string,
+        {
+          executor: (
+            params: { connector: Connector } | undefined
+          ) => Promise<void>
+          params?: { connector: Connector } | undefined
+        }
+      >
+    ) =>
+    (id: string) => {
+      switch (id) {
+        case SignInType.NearWalletSelector:
+          return executors[id].executor(undefined)
+        case SignInType.WalletConnect:
+          return executors[id].executor(executors[id].params)
+        default:
+          throw new Error("Invalid executor")
+      }
+    }
 
-    alert(`Switched account to ${nextAccountId}`)
-  }
+  useEffect(() => {
+    if (!accountId && !address) {
+      setState({
+        address: undefined,
+        network: undefined,
+        keyStore: undefined,
+      })
+    }
+    if (accountId) {
+      setState({
+        address: accountId,
+        network: "near:mainet",
+        keyStore: SignInType.NearWalletSelector,
+      })
+    }
+    if (address) {
+      setState({
+        address,
+        network: chain?.id?.toString() ?? "unknown",
+        keyStore: SignInType.WalletConnect,
+      })
+    }
+  }, [accountId, address, chain])
 
   return {
-    handleSignIn,
-    handleSignOut,
-    handleSwitchWallet,
-    handleSwitchAccount,
+    async signIn({
+      id,
+      params,
+    }: {
+      id: SignInType
+      params: { connector: Connector } | undefined
+    }): Promise<void> {
+      await abstractExecutor({
+        [SignInType.NearWalletSelector]: {
+          executor: handleSignInViaNearWalletSelector,
+          params: undefined,
+        },
+        [SignInType.WalletConnect]: {
+          executor: handleSignInViaWalletConnect,
+          params,
+        },
+      })(id)
+    },
+
+    async signOut({ id }: { id: SignInType }): Promise<void> {
+      await abstractExecutor({
+        [SignInType.NearWalletSelector]: {
+          executor: handleSignOutViaNearWalletSelector,
+          params: undefined,
+        },
+        [SignInType.WalletConnect]: {
+          executor: handleSignOutViaWalletConnect,
+          params: undefined,
+        },
+      })(id)
+    },
+
+    // TODO: Implement this
+    signMessage: async ({
+      id,
+      message,
+    }: { id: SignInType; message: string }) => {},
+
+    sendTransaction: async ({
+      id,
+      type,
+      transactions,
+    }: {
+      id: SignInType
+      type: SendTransactionType
+      transactions: SignAndSendTransactionsParams
+    }) => {
+      await abstractExecutor({
+        [SignInType.NearWalletSelector]: {
+          executor: async () => {
+            if (type === SendTransactionType.SignAndSendTransactions) {
```

Comment:
Not sure why `SendTransactionType.SignAndSendTransactions` is needed, could you explain?